Objects/complexobject.c

```diff
@@ -53,11 +53,63 @@ _Py_c_neg(Py_complex a)
 }
 
 Py_complex
-_Py_c_prod(Py_complex a, Py_complex b)
+_Py_c_prod(Py_complex z, Py_complex w)
 {
-    Py_complex r;
-    r.real = a.real*b.real - a.imag*b.imag;
-    r.imag = a.real*b.imag + a.imag*b.real;
+    double a = z.real, b = z.imag, c = w.real, d = w.imag;
+    double ac = a*c, bd = b*d, ad = a*d, bc = b*c;
+    Py_complex r = {ac - bd, ad + bc};
+
+    /* Recover infinities that computed as nan+nanj.  See e.g. the C11,
+       Annex G.5.2, routine _Cmultd(). */
+    if (isnan(r.real) && isnan(r.imag)) {
```

Comment:
To remove one level of indentation, how about:

```C
if (!isnan(r.real) || !isnan(r.imag)) {
	return r; 
}

int recalc = 0;
```

Reply:
I don't think it's a good idea:
1. this will be not a _Cmultd() routine anymore
2. regular case (non-nans) will cost two isnan() checks, instead of just one.

Reply:
While 1 is not really an issue for me, you're right on 2 so let's keep what you wrote.